url4c41254592_topskychen_AuthTopk_tgz-pJ8-quickhull3d_QuickHull3DTestJ8/src/utility/geo/VoronoiNeighbors.java

```diff
@@ -0,0 +1,200 @@
+/**
+ * 
+ */
+package utility.geo;
+
+import java.io.BufferedInputStream;
+import java.io.BufferedOutputStream;
+import java.io.DataInputStream;
+import java.io.DataOutputStream;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileNotFoundException;
+import java.io.FileOutputStream;
+import java.io.FileReader;
+import java.io.IOException;
+import java.io.LineNumberReader;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.Queue;
+import java.util.Scanner;
+
+
+import mesh.Delaunay;
+
+import utility.Compare.buildBtreeOfPoints;
+import utility.security.DataIO;
+
+/**
+ * @author chenqian
+ *
+ */
+public class VoronoiNeighbors {
+
+	private ArrayList<float[]> points = null;
+	private ArrayList<Integer> idmap = null;
+	private Delaunay delaunay = null;
+	
+	public void generateVoronoiNeighbors(String sourceFileName, String destFileName){
+		String line = null;
+		LineNumberReader lr = null;
+		try {
+			lr = new LineNumberReader(new FileReader(sourceFileName));
+		} catch (FileNotFoundException e1) {
+			// TODO Auto-generated catch block
+			e1.printStackTrace();
+		}
+		points = new ArrayList<float[]>();
+		idmap = new ArrayList<Integer>();
+		try {
+			while((line = lr.readLine()) != null){
+				String[] num = line.split(" ");
+				idmap.add(Integer.parseInt(num[1]));
+				float[] point = new float[2];
+				point[0] = Float.parseFloat(num[2]);
+				point[1] = Float.parseFloat(num[3]);
+				if(buildBtreeOfPoints.ISCAR){
+					point[0] = (int)(point[0] * buildBtreeOfPoints.SCALE);
+					point[1] = (int)(point[1] * buildBtreeOfPoints.SCALE);
+				}
+				points.add(point);
+				//if(DEBUG) System.err.println();
+			}
+		} catch (IOException e) {
+			// TODO Auto-generated catch block
+			e.printStackTrace();
+		}
+		delaunay = new Delaunay(points.toArray(new float[0][0]));
+		DataOutputStream dos = null;
+		try {
+			if(destFileName.endsWith(".nb") == false)destFileName += ".nb";
+			dos = new DataOutputStream(new BufferedOutputStream(new FileOutputStream(new File(destFileName))));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the resource created by the following constructor: *BufferedOutputStream*. The resource is referenced at lines 71, 74, 75, 83, 91, 101, 104. The resource is closed at line 105. There are other execution paths that do not contain closure statements, for example, when DataOutputStream.flush throws an exception. Ensure the object returned by *BufferedOutputStream()* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).